scheds/rust/scx_rustland/src/bpf/main.bpf.c

```diff
@@ -311,18 +314,26 @@ static void dispatch_task(struct task_struct *p, u64 dsq_id, u64 enq_flags)
 		 * in the meantime), we can simply ignore the task, as it has
 		 * been dequeued and re-enqueued, so it will pick a valid CPU.
 		 */
+		scx_bpf_dispatch(p, dsq_id, slice, enq_flags);
+
 		cpu = dsq_to_cpu(dsq_id);
 		if (!bpf_cpumask_test_cpu(cpu, p->cpus_ptr)) {
 			cpu = scx_bpf_task_cpu(p);
 			if (!bpf_cpumask_test_cpu(cpu, p->cpus_ptr))
-				return;
-			dsq_id = dsq_to_cpu(cpu);
+				dsq_id = SHARED_DSQ;
+			else
+				dsq_id = cpu_to_dsq(cpu);
```

Comment:
In this path, it probably should always dispatch to `SHARED_DSQ` because there's no guarantee that there's going to be another racing `set_cpumask()` betwen the second `bpf_cpumask_test_cpu()` and the cancel/dispatch. That said, do you still see stalls if you simply do cancel as follows?
```
scx_bpf_dispatch(p, dsq_id, slice, enq_flags);
if (!bpf_cpumask_test_cpu(dsq_to_cpu(dsq_id), p->cpus_ptr))
        scx_bpf_dispatch_cancel();
```

Reply:
It'd be also useful to keep a counter for this event.

Reply:
> In this path, it probably should always dispatch to `SHARED_DSQ` because there's no guarantee that there's going to be another racing `set_cpumask()` betwen the second `bpf_cpumask_test_cpu()` and the cancel/dispatch. That said, do you still see stalls if you simply do cancel as follows?
> ```
> scx_bpf_dispatch(p, dsq_id, slice, enq_flags);
> if (!bpf_cpumask_test_cpu(dsq_to_cpu(dsq_id), p->cpus_ptr))
>         scx_bpf_dispatch_cancel();
> ```

Hm.. that's true, I'm surprised that I haven't triggered this condition yet, even with massive stress-ng race-sched runs. But it definitely seems safer to just bounce the task to the shared DSQ. I'll change this.

About the stall yes, if I just return after cancel I can easily trigger stalls and in the trace I can see that the stalling task is not dispatched to any DSQ.

Reply:
> It'd be also useful to keep a counter for this event.

This is a good idea, I'll add it!

Reply:
> About the stall yes, if I just return after cancel I can easily trigger stalls and in the trace I can see that the stalling task is not dispatched to any DSQ.

Hmm... that shouldn't be the case because the task gets requeued after cpumask is updated and thus should re-dispatch the task. Once you land this PR, I'll test and see what's going on.